File: 模板/python/Comb-Perm.py
MOD = 10**9+7
N = 10**6+10

fac = [1] * N
inv = [1] * N
for i in range(1, N):
    fac[i] = fac[i-1] * i % MOD
for i in range(2, N):
    inv[i] = (MOD - MOD // i) * inv[MOD % i] % MOD
for i in range(1, N):
    inv[i] = (inv[i-1] * inv[i]) % MOD

def C(n, m):
    return fac[n] * inv[m] * inv[n-m] % MOD if n >= m and m >= 0 else 0
def P(n, m):
    return fac[n] * inv[n-m] % MOD if n >= m and m >= 0 else 0

def lucas(n, m):
    if n < m:
        return 0
    elif n < MOD:
        return C(n, m)
    else:
        return lucas(n // MOD, m // MOD) * lucas(n % MOD, m % MOD) % MOD

File: 模板/cpp/Comb-Perm.cpp
ll fac[N];
ll inv[N];

void init(ll mod) {
    fac[0] = 1;
    for (int i = 1; i < N; i++) {
        fac[i] = (fac[i - 1] * i) % mod;
    }
    inv[0] = inv[1] = 1;
    for (int i = 2; i < N; i++) {
        inv[i] = (ll) (mod - mod / i) * inv[mod % i] % mod;
    }
    for (int i = 1; i < N; i++) {
        inv[i] = (inv[i - 1] * inv[i]) % mod;
    }
}

ll C(ll n, ll m, ll mod) {
    return n < m ? 0 : fac[n] * inv[m] % mod * inv[n - m] % mod;
}

ll P(ll n, ll m, ll mod) {
    return n < m ? 0 : fac[n] * inv[n - m] % mod;
}

ll lucas(ll n, ll m, ll mod)
{
    if(n < m) return 0;
    else if(n < mod) return C(n, m, mod);
    else return lucas(n / mod, m / mod, mod) * lucas(n % mod, m % mod, mod) % mod;
}
